X509: Add support for directly checking leaf cert EKU

DDvO opened the pull request:
This aims at adding simple support for conveniently checking allowed Extended Key Usage 
specified by any OID, not only the well-known ones defined in RFC 5280 etc. and explicitly supported by OpenSSL.
The need for this was mentioned by https://github.com/openssl/openssl/issues/13965#issuecomment-776249427.

So far, EKUs could only be checked rather indirectly via giving a purpose, which is more high-level concept that is implemented in a rather cumbersome and hard-to-extend way.

The new solution is implemented as follows.
* `X509_VERIFY_PARAM` is extented by a new field:
   `int eku;                    /* Extended Key Usage NID to check */`
   and a new flag: `X509_V_FLAG_EKU_INCLUDED`
* a couple of new functions to set the required EKU at various levels:
  `int X509_VERIFY_PARAM_set_eku(X509_VERIFY_PARAM *param, int eku)`
  `int SSL_CTX_set_eku(SSL_CTX *s, int eku)`
  `int SSL_set_eku(SSL *s, int eku)`
* new CLI options `-eku` and `-eku_included`
* a new function `int X509_check_eku(X509 *x, int nid, int eku_included)` 
  to check if the given EKU OID is allowed by the cert, called by `check_chain()` for leaf certs

The  semantics is that if a target EKU is specified in the verification parameters,
the leaf cert (which may be an EE or CA cert) must be compatible with the specified EKU, 
overriding any purpose checks set in the verification parameters.
If the `X509_V_FLAG_EKU_INCLUDED` verification flag is set the EKU extension
must be present in the certificate and must include the specified EKU.

Here is a simple example use:
```
apps/openssl s_client -connect login.live.com:443 -verifyCAfile \
DigiCertGlobalRootCA.crt.pem -verify_return_error -eku serverAuth
```

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

richsalz commented:
As I have written elsewhere in a comment, this behavior is problematic.  Microsoft, for example, requires all intermediate CA's to have all the EKU values for anything they sign.

DDvO commented:
This should be possible to get accommodated in a straightforward way 
when there is more clarity about this.

richsalz commented:
There *is* clarity.  Real-world WebPKI chains have both EKU in the intermediate CA's and EKU not in the intermediates.
See P47 of https://cabforum.org/wp-content/uploads/CA-Browser-Forum-BR-1.6.3.pdf

DDvO commented:
> There _is_ clarity. Real-world WebPKI chains have both EKU in the intermediate CA's and EKU not in the intermediates.

Well, the whole topic is not as universally clear as one might think/hope.

First of all, you speak of "intermediate CAs", while CA/B Forum speaks of "subordinate CAs", and there is still unclarity already about their exact meaning - see also the recent discussion on the LAMPS WG mailing list.
Hopefully we can simply say here: There are (typically self-signed) root CA certs, end-entity certs, and any certs in between are intermediate/subordinate CA certs (with some of them also known as issuing CA certs).

DDvO commented:
> See P47 of https://cabforum.org/wp-content/uploads/CA-Browser-Forum-BR-1.6.3.pdf

Thanks for this reference. The page numbers in that PDF document are inconsistent with those in its index. I assume you meant the number 47 in the index, which translates to page 41, where the section 7.1.2. "Certificate Content and Extensions; Application of RFC 5280" starts.

According to this,
> for Root CA Certificates, extendedKeyUsage MUST NOT be present.

So I think we are on the same page that no EKU is to be checked here because there is none.

On the following page it says :
> extkeyUsage (optional)
> For Subordinate CA Certificates to be Technically constrained in line with section 7.1.5, then either the value id-kp-serverAuth [RFC5280] or id-kp-clientAuth [RFC5280] or both values MUST be present**.
> ** Generally Extended Key Usage will only appear within end entity certificates (as highlighted in RFC 5280 (4.2.1.12)), however, Subordinate CAs MAY include the extension to further protect relying parties until the use of the extension is consistent between Application Software Suppliers whose software is used by a substantial portion of Relying Parties worldwide.

So in case the EKU extension is present in non-root CA certs, we likely should check that it allows for the same EKU that the leaf/EE cert is to be checked for. Something like this is already done by OpenSSL for the so-called purposes.

DDvO commented:
I've just extended the documentation
and extended the scope of the direct EKU checking to all untrusted certs (including the leaf/target cert) in analogy to the classic 'purpose'-oriented checks.
This should fulfill also the requirements of the CA/Browser forum cited above.

t8m commented:
IMO this is clearly Post 3.0.0 material.

DDvO commented:
Fine by me, so removed the (tentative) milestone.

richsalz commented:
> First of all, you speak of "intermediate CAs", while CA/B Forum speaks of "subordinate CAs", and there is still unclarity already about their exact meaning - see also the recent discussion on the LAMPS WG mailing list.

You have to evaluate terms in their context.  LAMPS is not the CA/B Forum.

I suggest you read all of the BR's before submitting this.

DDvO commented:
The CAB BR document https://cabforum.org/wp-content/uploads/CA-Browser-Forum-BR-1.7.3.pdf
is newer and does not have the line number discrepancies.
Which section(s) do you see relevant for our topic here?
I suppose "7.1.2 Certificate Content and Extensions; Application of RFC 5280".

richsalz commented:
I suggest you read the whole thing and RFC 5280. I'm not going to give subsections; if you're proposing to change things you should have a good background.

DDvO commented:
Added respective test cases and slightly refined the implementation.

DDvO commented:
> IMO this is clearly Post 3.0.0 material.

@t8m, why do you think so?
This new feature is pretty straightforward and 
is 100% backward compatible: it not explicitly used, everything stays as it was.

t8m commented:
> > IMO this is clearly Post 3.0.0 material.
> 
> @t8m, why do you think so?
> 
> This new feature is pretty straightforward and
> is 100% backward compatible: if not explicitly used, everything stays as it was.

Didn't the PR morph a little bit since? Maybe it is now acceptable if someone approves. I am not reviewing this now though as I am busy with other things and I do not feel like this is my particular area of expertise.

richsalz commented:
The code in verify options should be refactored to an `opt_oid` function like `opt_md` etc. The void cast at line 600 goes against common style; use `__owur` for places where you *can't* ignore the return value.

The implementation requires users to call a utility function to convert an OID to a NID, the manpages needs a link.

There's a limit of only one OID which seems limiting; "I want all of oid1 oid2" or "I want one of oid1 oid2" both seem like reasonable use-cases. To me the discussion isn't complete yet.  Do we have experience or real-world examples to help tip the scales? To me, this is the biggest reason why this PR isn't applicable to 3.0

Any PR which changes more than two dozen files is a stretch to call straightforward :)

t8m commented:
> Any PR which changes more than two dozen files is a stretch to call straightforward :)

Totally agree!

DDvO commented:
> The code in verify options should be refactored to an `opt_oid` function like `opt_md` etc. The void cast at line 600 goes against common style; use `__owur` for places where you _can't_ ignore the return value.

Ok, I've adapted `apps/lib/opt.c` accordingly.

> The implementation requires users to call a utility function to convert an OID to a NID, the manpages needs a link.

So added a hint on using `OBJ_txt2nid()` 
to `SSL_CTX_set_eku.pod` and `X509_VERIFY_PARAM_set_flags.pod`.

DDvO commented:
> Any PR which changes more than two dozen files is a stretch to call straightforward :)

I did not write 'trivial' but 'pretty straightforward' :wink:

On closer look one can easily see that most of the files adapted only contain superficial stuff such as CLI integration, tests, and documentation. 
In total these might be seen as a little bulky, but not difficult.
Recently I had included some more general improvements that I came across while preparing this PR; I've just moved those to #14413 for making reviewing easier.

Actually the core of this PR is just a single function added to `v3_purp.c`:
```
/* Directly check for allowed extended key usage (EKU) */
int X509_check_eku(X509 *x, int nid)
{
    int eku, i, n = sk_ASN1_OBJECT_num(x->ex_ekus);

    if (nid == NID_anyExtendedKeyUsage || n <= 0)
        return 1;
    for (i = 0; i < n; i++) {
        eku = OBJ_obj2nid(sk_ASN1_OBJECT_value(x->ex_ekus, i));
        if (eku == NID_anyExtendedKeyUsage || eku == nid)
            return 1;
    }
    return 0;
}
```
and an addition to `check_extensions()` in `x509_vfy.c` calling it:
```
        if (eku != 0 && i < ctx->num_untrusted)
            /* EKU requirement overrides purpose check for untrusted certs */
            CB_FAIL_IF(X509_check_eku(x, eku) <= 0, ctx, x, i,
                       X509_V_ERR_INVALID_EXTENDED_KEY_USAGE);
```
 which I do consider straightforward.

DDvO commented:
> There's a limit of only one OID which seems limiting; "I want all of oid1 oid2" or "I want one of oid1 oid2" both seem like reasonable use-cases.

I was contemplating that, too, but do not see much point in the possibility to specify more that one EKU OID - do you have a concrete example where this would be useful?

Note that also the closely related purpose requirement only allows to specify exactly one purpose.

For both the new EKU requirement option and the existing purpose requirement option, 
if one really wants/needs to check for a combination of values, 
one can achieve this simply by calling the verification for each of them and 
then combine the results accordingly.

So in my view allowing to require more than one EKU at a time would just complicate the expected typical use and the implementation.

DDvO commented:
> To me the discussion isn't complete yet. Do we have experience or real-world examples to help tip the scales? To me, this is the biggest reason why this PR isn't applicable to 3.0
 
Good question - so far I only see pretty simple use cases, 
but likey others have more experience here - @kaduk ?

richsalz commented:
> Good question - so far I only see pretty simple use cases,

Have you looked at IPsec?  If not, you should.  (And no, I don't have a summary/reading list.)

DDvO commented:
> > Good question - so far I only see pretty simple use cases,
> 
> Have you looked at IPsec? If not, you should. (And no, I don't have a summary/reading list.)

@richsalz, I'm doing this and many other cert-related contributions in my spare time,
aiming at helping out the project and the global community on a goodwill basis.
I cannot afford the extra time and burden to do extensive studies etc. just for that.

richsalz commented:
I understand that you are doing things on your spare time and thanks.

But if you are doing things that don't actually help, based on your limited knowledge and understanding, then it is hurtful to the project.  Maybe give the whole certificate validation stuff a break for now until someone has time to understand how things are used in the real world.

DDvO commented:
@richsalz, your comments in particular this thread sound like you believe you have 
superior knowledge, understanding, and experience. No problem with that, but if so, 
why aren't you more constructive and share whatever portion of it is relevant here?

richsalz commented:
>your comments in particular this thread sound like you believe you have superior knowledge, understanding, and experience. 

- I know what I don't know.
- Previous comments I've made on PR's about certificate and chain validation based on my experience have gone unheeded.
- My time and resources are mostly about getting 3.0 released.

DDvO commented:
Keeping the status quo of certificate (chain) validation in the sense that 
no regression should creep in is certainly an aim that all of us agree on.
I also see the value in aiming to get the next release out as soon as reasonably possible,
and this calls for minimizing activity on this topic.

On the other hand, the more I get into the topic the more technical debt I find - 
see in particular the meta issue #13780 I brought up early this year.
I believe it is valuable to fix some portion of that before the release,
at least the correctness and completeness of the documentation.

Feature extensions of the cert validation like the one proposed in this PR
is something that appears less urgent, at least to me. 
My aim here, as mentioned, was/is to help @kaduk and the user base 
he is supporting, while I do not know how much urgency they see in this.

So far, no milestone has been placed on this PR.
What do others think on whether it's better to further pursue this PR now or post-3.0?

t8m commented:
I still prefer Post 3.0.0 on this one.

DDvO commented:
@kaduk, are you fine with marking this post-3.0.0?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 371 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 402 days ago

DDvO commented:
Rebased to fix merge conflicts.

> I still prefer Post 3.0.0 on this one.

Since meanwhile we are well past 3.0.0, please let's resume on this PR.

vdukhovni commented:
Note, any purpose/EKU checks need to apply equally to both "trusted" and "untrusted" certificates. The same (intermediate) certificate may appear either in the chain presented by the peer or found in the trust store, and there is no good reason to apply different policy to its meaning.

Furthermore, OpenSSL supports "TRUSTED CERTIFICATE" objects with externally designated EKUs applied by local policy, rather than an issuing CA.  These especially should be honoured, despite the certificate being "trusted" (for the indicated purpose).

DDvO commented:
Thanks @vdukhovni for your comments.

> Note, any purpose/EKU checks need to apply equally to both "trusted" and "untrusted" certificates. The same (intermediate) certificate may appear either in the chain presented by the peer or found in the trust store, and there is no good reason to apply different policy to its meaning.

Originally, this PR aimed at doing the new EKU check only on leaf certs,
but due to a reference to CA/B Forum requirements this was extended to checking also certs along the chain.
From your remark I conclude that checking just untrusted certs in the chain is  inconsistent.
So a chain-oriented check should be done for all certs in the chain (excluding the trust anchor),
regardless whether they are locally considered trusted or untrusted right?

OTOH, there are requirements such those in https://datatracker.ietf.org/doc/html/draft-ietf-lamps-cmp-updates#section-2.2 that pertain to the leaf cert only:
```
The meaning of the id-kp-cmKGA EKU is as follows:

   CMP KGA:  CMP Key Generation Authorities are identified by the id-kp-
             cmKGA extended key usage.  The CMP KGA knows the private
             key it generated on behalf of the end entity.  This is a
             very sensitive service and therefore needs specific
             authorization.  This authorization is with the CA
             certificate itself.  Alternatively, the CA MAY delegate the
             authorization by placing the id-kp-cmKGA extended key usage
             in the certificate used to authenticate the origin of the
             generated private key or the delegation MAY be determined
             through local configuration of the end entity.
```
To support such scenarios, I meanwhile tend to go back again to doing the new EKU check
only on leaf certs (regardless whether they are trusted or untrusted, of course).

vdukhovni commented:
> From your remark I conclude that checking just untrusted certs in the chain is inconsistent. So a chain-oriented check should be done for all certs in the chain (excluding the trust anchor), regardless whether they are locally considered trusted or untrusted right?

Correct, all the certificates in the chain up to and *including* the trust anchor, where any self-signed certificate with no explicit EKUs is assumed to be valid for all "purposes", but if it has auxiliary trust settings (restricting what it is trusted for) then these also apply...

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 278 days ago

vdukhovni commented on apps/lib/opt.c:
```diff
@@ -480,6 +478,16 @@ int opt_check_md(const char *name)
+{
+    int nid = OBJ_txt2nid(name);
+
+    if (nid == 0)
```
Should this be `0` or `NID_undef`?

vdukhovni commented on apps/lib/opt.c:
```diff
@@ -738,6 +746,11 @@ int opt_verify(int opt, X509_VERIFY_PARAM *vpm)
         }
         break;
+    case OPT_V_EKU:
+        if ((i = opt_oid(opt_arg(), "Extended Key Usage")) == 0)
+            return 0;
+        X509_VERIFY_PARAM_set_eku(vpm, i);
```
Variable names are cheap:
```suggestion
        if ((eku_nid = opt_oid(opt_arg(), "Extended Key Usage")) == NID_undef)
            return 0;
        X509_VERIFY_PARAM_set_eku(vpm, eku_nid);
```

vdukhovni commented on crypto/x509/x509_vfy.c:
```diff
@@ -455,14 +457,16 @@ static int check_purpose(X509_STORE_CTX *ctx, X509 *x, int purpose, int depth,
 /*-
- * Check extensions of a cert chain for consistency with the supplied purpose.
+ * Check extensions of a cert chain for consistency with the supplied purpose
+ * and leaf cert EKU.
```
```suggestion
 * or instead any explicitly specified EKU.
```

vdukhovni commented on crypto/x509/x509_vfy.c:
```diff
@@ -590,8 +594,12 @@ static int check_extensions(X509_STORE_CTX *ctx)
             }
         }
 
+        if (eku != 0 && i < ctx->num_untrusted)
```
```suggestion
        if (eku != NID_undef)
```
Check the whole chain, not just the untrusted certificates.

vdukhovni commented on crypto/x509/x509_vpm.c:
```diff
@@ -507,6 +515,7 @@ static const X509_VERIFY_PARAM default_table[] = {
      0,                         /* check time to use */
      0,                         /* inheritance flags */
      X509_V_FLAG_TRUSTED_FIRST, /* flags */
+     0,                         /* eku */
```
```suggestion
     NID_undef,                         /* eku */
```
And similar cases below.

vdukhovni commented on doc/man1/openssl-verification-options.pod:
```diff
@@ -151,12 +151,13 @@ The first step is to check that each certificate is well-formed.
-If the B<-purpose> option is not given then no such checks are done
-except for SSL/TLS connection setup,
+for consistency with the supplied purpose or Extended Key Usage (EKU),
+where the latter overrides any purpose requirements for untrusted certificates.
```
Not just *untrusted certificates*, the entire chain.

vdukhovni commented on doc/man3/SSL_CTX_set_eku.pod:
```diff
@@ -0,0 +1,49 @@
+This must be an NID corresponding to an EKU OID, for example B<NID_server_auth>,
+which may have been obtained by parsing a text string using L<OBJ_txt2nid(3)>.
+It determines the acceptable EKU for the peer certificate and its chain,
+If provided, it overrides any purpose requirements for untrusted certificates.
```
Not just untrusted.  Explain that setting this to `NID_undef` resets the check.

vdukhovni commented on doc/man3/X509_VERIFY_PARAM_set_flags.pod:
```diff
@@ -87,6 +89,12 @@ See the B<INHERITANCE FLAGS> section for a description of these bits.
+This must be an NID corresponding to an EKU OID, for example B<NID_server_auth>,
+which may have been obtained by parsing a text string using L<OBJ_txt2nid(3)>.
+It determines the acceptable EKU for the certificate chain to be verified.
+If provided, it overrides any purpose requirements for untrusted certificates.
```
Not just untrusted.

DDvO commented:
Thank you @vdukhovni for taking the initiative to bring this PR forward again.

I've just rebased it and followed all your above new reviewing comments.

DDvO commented:
I've just adapted the title and the commit message of this PR to reflect the new semantics,
namely to affect not only leaf certs or untrusted certs, but the entire chain.

DDvO commented on doc/man3/SSL_CTX_set_eku.pod:
```diff
@@ -0,0 +1,50 @@
+this must be an NID corresponding to an EKU OID, for example B<NID_server_auth>,
+which may have been obtained by parsing a text string using L<OBJ_txt2nid(3)>.
+If provided, the NID determines the acceptable EKU for the peer certificate
+and its chain, overriding any purpose requirements for the entire chain.
```
Done both.

DDvO commented on doc/man3/X509_VERIFY_PARAM_set_flags.pod:
```diff
@@ -87,6 +89,13 @@ See the B<INHERITANCE FLAGS> section for a description of these bits.
+this must be an NID corresponding to an EKU OID, for example B<NID_server_auth>,
+which may have been obtained by parsing a text string using L<OBJ_txt2nid(3)>.
+If provided, the NID determines the acceptable EKU for the certificate chain
+to be verified, overriding any purpose requirements for the entire chain.
```
Done here as well.

DDvO commented:
In addition to handling the above new review comments,
I renamed the new `ERR_INVALID_EXTENDED_KEY_USAGE` to the more to-the-point `X509_V_ERR_WRONG_EXTENDED_KEY_USAGE`.

Doing so, I found that the order of many error texts listed in `doc/man3/X509_STORE_CTX_get_error.pod` 
were not consistent with order `include/openssl/x509_vfy.h.in` and `crypto/x509/x509_txt.c`.
Fixed this, and on this occasion added a couple of missing entries 
and added a comment in  `crypto/x509/x509_txt.c` that the three files must be kept in sync.

DDvO commented:
I noticed that a few more texts had to be updated according to the extended semantics checking the EKU of all the chain.
Fixed this, and on this occasion also improved the overall description of certificate validation in `openssl-verification-options.pod`.

DDvO commented:
I'm going to backport to 3.0 and 3.1 the general doc improvements in `openssl-verification-options.pod` and `X509_STORE_CTX_get_error.pod` that I added here yesterday and today.

vdukhovni commented on apps/include/opt.h:
```diff
@@ -38,6 +39,8 @@
         { "policy", OPT_V_POLICY, 's', "adds policy to the acceptable policy set"}, \
         { "purpose", OPT_V_PURPOSE, 's', \
             "certificate chain purpose"}, \
+        { "eku", OPT_V_EKU, 's', \
```
Perhaps the option name should be a bit more descriptive:
```suggestion
        { "required-eku", OPT_V_EKU, 's', \
```
If you disagree, I won't insist.

DDvO replied:
Well, most of the OpenSSL CLI option names are pretty short and not fully self-explanatory.
I believe that in the given context `-eku` should be clear enough, and it is fully in line with `-purpose`.

(And if we go for a longer name like the one suggested above, for consistency with others it should involve a `_` rather than a `-`.)

vdukhovni commented on crypto/x509/x509_vpm.c:
```diff
@@ -167,6 +167,7 @@ int X509_VERIFY_PARAM_inherit(X509_VERIFY_PARAM *dest,
     to_default = (inh_flags & X509_VP_FLAG_DEFAULT) != 0;
     to_overwrite = (inh_flags & X509_VP_FLAG_OVERWRITE) != 0;
 
+    x509_verify_param_copy(eku, 0);
```
```suggestion
    x509_verify_param_copy(eku, NID_undef);
```

DDvO replied:
Done

vdukhovni commented on doc/man1/openssl-verification-options.pod:
```diff
@@ -13,19 +13,22 @@ I<command>
+trust anchor, which usually is a certificate that due to some policy is trusted.
+Certificate validation can be done relative to a given Extended Key Usage (EKU)
+or a given I<purpose>, which is the intended use of the target certificate,
+such as SSL server, or (by default) for any EKU and any purpose.
```
The above and other editorial changes, unrelated to support for EKU constraint enforcement belong in a separate PR, please split off.

DDvO replied:
As requested, I removed those (not directly related) changes from this PR
though this will cause merge conflicts when the separate PR(s) will get merged.

DDvO replied:
I found that I already had a PR, namely #18764, where those extra changes to `openssl-verification-options.pod` fit perfectly, so moved them there.

vdukhovni replied:
> As requested, I removed those (not directly related) changes from this PR though this will cause merge conflicts when the separate PR(s) will get merged.

You can use the current PR as the base revision for the separate PR and indicate that its review should wait for the current one to merged, or just wait to submit it. Whatever works for you.

DDvO replied:
Thank you for these hints.
At that point I had already split the commits to separate PRs,
and as the changes  to `openssl-verification-options.pod` done meanwhile in this PR are confined to a single paragraph, the merge conflicts with #18764 will be bearable.

DDvO commented:
@vdukhovni can you approve now?

wbl commented:
I have some serious doubts about the utility and wisdom of this patch.

First off we don't have any concrete usecase here. It's not like the national turboencabulator society said "this is the semantics of our turboencabulator PKI that all turboencabulators get certificates from", or applications needed this and are saying so.

Secondly this patch would impose a particular set of semantics on EKU usage across all PKIs: the most convenient to implement would be what this patch wants, which is the EKU must be on all certs in the chain if there are any EKUs used, and we couldn't easily add other behaviors in a compatible way.

I think EKU limitation is a good goal, and there aren't good ways to do it now. But this needs to be something in LAMPS or some other fora where people can work through the issues. Not a unilateral change that makes only one possibility easy to use, without any discussion of the PKI issues.

DDvO commented:
Thank you for your comment.

As mentioned, this PR was motivated by a comment by @kaduk: https://github.com/openssl/openssl/issues/13965#issuecomment-776249427

At first it was only checking the EKU extension in the leaf cert,
and due to comments by @richsalz and @vdukhovni it was extended to check the whole chain.

I like the idea of bringing this approach to the LAMPS WG for discussion.
What do others think?

richsalz commented:
> due to comments by @richsalz and @vdukhovni it was extended to check the whole chain.

Not sure about Viktor, but it is incorrect to say that I wanted the whole chain. (Maybe that isn't what you intended to imply.) I was just pointing out that, at the time -- *two years ago* -- this wasn't compliant with some standards.

I share Watson's concerns and suggestions.

vdukhovni commented:
> Thank you for your comment.
> 
> As mentioned, this PR was motivated by a comment by @kaduk: [#13965 (comment)](https://github.com/openssl/openssl/issues/13965#issuecomment-776249427)
> 
> At first it was only checking the EKU extension in the leaf cert, and due to comments by @richsalz and @vdukhovni it was extended to check the whole chain.
> 
> I like the idea of bringing this approach to the LAMPS WG for discussion. What do others think?

Note that by the time I reviewed the PR, it was **not** looking at **just** the leaf cert, it was looking at all *untrusted* intermediate CA certs also.  My objection was to that aspect of the PR.  If it is going to restrict *some* CAs, it needs to restrict all CAs.  If the feature was meant to be EE-only, that's not how I found it.

DDvO commented:
Hmm, maybe I misunderstood various early comments on this PR,
but if so, I wonder why this was not discovered until yesterday, over two years later.
Of course, relevant standards may have evolved sine then.

Anyway, we can still steer the implementation in this PR in the most suitable direction;
technically any such adaption will be straightforward to do.

DDvO commented:
What exact EKU checks should be done on (intermediate, or possibly leaf) CA certs appears debatable.
Part of the reason likely being that [RFC 5280 says](https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.12):
> In general, this extension will appear only in end entity certificates

OpenSSL since long has the related 'purpose' concept, which is 
* non-trivial
* hard to extend
* still not fully documented (where since many years I am the only one who cares to provide improvements on this, see for instance #18764), 
* hardly known to, let alone understood by, average OpenSSL users
* I bet, very rarely used explicitly even by expert OpenSSL users.

Note that of  the purpose checkers built in `v3_purp.c`, 
the ones for or SSL client, SSL server and S/MIME purposes check the EKU extensions along the whole chain
while all others do any EKU checks only for the leaf cert, while not distinguishing between EE and CA certs there.
I doubt that this is due to deliberate design and careful implementation - 
to me this rather feels like accidental inconsistency that went unnoticed 
because most CA certs do not have EKU restrictions in place.

Given the status quo and its history, maybe better called legacy,
one may wonder why suddenly people start raising/confirming concerns on EKU checking on CA certs on this PR.

vdukhovni commented:
1. It should not matter whether the leaf cert is or isn't tagged as a CA. Self-signed certs are often tagged as CAs, just because.
2. Indeed be clear on what use-cases this PR is for, and what is expected/useful in these cases.
3. I am not convinced that it makes sense to require some default EKU in the issuer certificates that is different from the one required in the leaf cert. If the issuer EKU checks are to be performed (when an override leaf EKU is specified) they should probably use the same EKU. Otherwise, perhaps they should not be performed, and the user can then use this mechanism to get the strict RFC5280 EKU semantics (opting out of local policy restricting or conversely promoting issuer CA trust).
4. Note that when a certificate in the trust store is NOT self-signed, it needs explicit (external "trusted certificate") EKUs to be a trust anchor. The default "fit for all purposes" trust-anchor status applies only to self-signed CAs, and originally intended as a "legacy" trust mechanism, with explicit tagging as a "trusted certificate" the expected trust anchor format, but that's not how things worked out in practice.  The trust-anchor "CA bundles" shipped with OpenSSL don't bother to take the high road and wrap the CAs as "trusted certificates".
5. The more fine-grained trust-anchor management is found in the NSS cert store, the MacOS trust store, the Microsoft trust store, ...  It just never took off in the OpenSSL ecosystem.

DDvO commented:
> 1. It should not matter whether the leaf cert is or isn't tagged as a CA. Self-signed certs are often tagged as CAs, just because.

Hmm, but setting the CA flag for self-signed EE certs is pretty misleading, if not dangerous.
I know that some folks still do so because of their own wrong understanding or as a workaround for getting along with unclean crypto implementations that for some reason insist/presume that self-signed certs must be CA certs.

This point explains just part of the inconsistencies in the OpenSSL purpose checking that I mentioned above.
In particular, to cover the situation of self-signed EE certs marked as CA, it would be sufficient to treat leaf CA certs equally to EE certs only if they are self-signed.

And whenever for any reasons OpenSSL treats in certain situations EE certs as CA certs or vice versa, 
this should be clearly documented. 
So far, almost no-one is aware of all those subtleties in cert (chain) validation.

> 2. Indeed be clear on what use-cases this PR is for, and what is expected/useful in these cases.

This PR was originally pretty clear, but then its semantics got blurred/distracted by the initial comments I received.
Now I tend to go back more or less to the original semantics (where any EKU extensions of non-leaf certs were ignored).
Still I would like to hear comments on our recent exchange from @kaduk, who inspired this PR in https://github.com/openssl/openssl/issues/13965#issuecomment-776249427.

> 3. I am not convinced that it makes sense to require some default EKU in the issuer certificates that is different from the one required in the leaf cert. If the issuer EKU checks are to be performed (when an override leaf EKU is specified) they should probably use the same EKU. Otherwise, perhaps they should not be performed, and the user can then use this mechanism to get the strict RFC 5280 EKU semantics (opting out of local policy restricting or conversely promoting issuer CA trust).

If I understand which point you want to make here, it seems that you advise not to let any (additional) purpose checking happen in case the new direct EKU check is done because this may lead to checking different EKUs for the intermediate certs than for the leaf cert. I agree and will adapt my most recent proposal accordingly.

> 4. Note that when a certificate in the trust store is NOT self-signed, it needs explicit (external "trusted certificate") EKUs to be a trust anchor. The default "fit for all purposes" trust-anchor status applies only to self-signed CAs, and originally intended as a "legacy" trust mechanism, with explicit tagging as a "trusted certificate" the expected trust anchor format, but that's not how things worked out in practice.  The trust-anchor "CA bundles" shipped with OpenSSL don't bother to take the high road and wrap the CAs as "trusted certificates".

Yes - it had taken me a while to learn about that OpenSSL "trusted certificate" mechanism, partly because of not very prominent documentation and because it is being generally neglected in practice.
BTW, also the `-partial_chain` option lets users get away without providing explicit trust attributes to certs in the trust store.

> 5. The more fine-grained trust-anchor management is found in the NSS cert store, the MacOS trust store, the Microsoft trust store, ...  It just never took off in the OpenSSL ecosystem.

Regardless which TA management mechanism is offered, it's hard to educate people to be sufficiently careful and precise on which TA they trust for which purposes, and 
even more advanced:  under which circumstances it is justified to gain trust via an existing TA into a further one, 
for instance using the `/cacerts` mechanism of EST and similar mechanisms available for CMPv2 and in CMP Updates.

vdukhovni commented:
You can define any reasonable semantics for your new option, based on real-world needs. Your constraints are:

* Do not change existing behaviour absent the new option.
* Do not distinguish between CA and not CA in EE certs
* Do not distinguish between "trusted' and "untrusted" intermediate CAs.  Any constraints applied to some CAs need to be applied to all CAs.

The rest is a question of a well-motivated use-case and accurate documentation.

DDvO commented:
> How about the following?
> 
> * Do the new optional EKU check only on EE certs - thus only on the leaf cert of a chain if it is not a CA cert
> * For backward compatibility, leave the (optional/default) purpose checks as they are, i.e., let them happen in addition
> 
> @kaduk would this be sufficient for your use cases?

In response to Viktor's most recent comments, I revise and extend this to:

* Add simple support for conveniently checking Extended Key Usage (EKU) in leaf certs during cert (chain) validation.
  This can be used, e.g., for checking the new EKU with `id-kp-cmKGA` defined by [CMP Updates](https://datatracker.ietf.org/doc/html/draft-ietf-lamps-cmp-updates#name-new-section-45-extended-key).
* The required EKU can be specified by any OID, not only the well-known ones defined in RFC 5280 etc. and explicitly supported by OpenSSL.
* Do the new optional check on leaf certs only, regardless whether the leaf cert has the CA tag or not.
The new EKU check is meant primarily for EE certs because EKU extensions are typically used only there.
* By default, let the check pass if the leaf cert does not contain an EKU extension.
   With an extra flag being set, such as `X509_V_FLAG_EKU_INCLUDED`, an  EKU extension MUST be present and include the required EKU OID. 
* When the EKU check is enabled, the (optional/default) 'classical' complex purpose checks are disabled.

DDvO commented:
I've just finished implementing the modified and extended semantics given above.

The new EKU check is now confined again to the leaf cert (which may be an EE or CA cert).
As before, the new check overrides any purpose checks, 
while all the existing options are not affected if the new option is not active.
A new subordinate option allows strictly requiring the presence of the specified EKU in the cert EKU extension.

Reviewing can commence.

richsalz commented:
I suggest you close this PR and open a new one with just the new changes and perhaps your last comment or two.

DDvO commented:
Rebased in the hope to get rid of unrelated `sslapitest` failure for `no-legacy` CI run.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 278 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 278 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 309 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 340 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 371 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 402 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 433 days ago

nhorman commented:
This issue has been idle for well over a year now.  While I understand there has been good work done here, in an effort to move our or backlog to a more reasonable state, I'm opting to close this.  If there is subsequent work to be done on it, or the poster wishes to pick it back up, please feel free to reopen it .

t8m commented:
This needs to be rebased and reviewed.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

t8m commented:
Could you please rebase the PR to resolve conflicts?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

DDvO commented:
> Could you please rebase the PR to resolve conflicts?

@t8m sorry that I again forgot to do. Done now.

t8m commented:
We should really consider this for 3.5. Needs a rebase.

DDvO commented:
Rebased again.
Ready for review.

t8m reviewed (approved):
LGTM, but I would definitely want to see a review from @vdukhovni

DDvO commented:
rebased to fix merge conflict

DDvO commented:
@vdukhovni can you please review this again?

DDvO commented:
Ping @vdukhovni

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 278 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 340 days ago
